crates/holochain/src/conductor/interface/websocket.rs

```diff
@@ -181,11 +191,89 @@ async fn recv_incoming_admin_msgs<A: InterfaceApi>(api: A, rx_from_iface: Websoc
         .await;
 }
 
+/// Takes an open connection and waits for an authentication message to complete the connection
+/// registration. If the connection is not authenticated within 10s or any other content is sent
+/// then the connection is dropped.
+///
+/// If the authentication succeeds then message handling tasks are spawned to handle normal
+/// communication with the client.
+fn authenticate_incoming_app_connection<A: InterfaceApi<Auth = AppAuthentication>>(
+    task_list: TaskListInner,
+    api: A,
+    mut rx_from_iface: WebsocketReceiver,
+    rx_from_cell: broadcast::Receiver<Signal>,
+    tx_to_iface: WebsocketSender,
+    installed_app_id: Option<InstalledAppId>,
+    port: u16,
+) {
+    let join_handle = tokio::task::spawn({
+        let task_list = task_list.clone();
+        async move {
+            let auth_payload_result = tokio::time::timeout(std::time::Duration::from_secs(10), async {
+                if let Ok(msg) = rx_from_iface.recv::<A::ApiRequest>().await {
+                    return match msg {
+                        ReceiveMessage::Authenticate(auth_payload) => {
+                            Ok(auth_payload)
+                        }
+                        _ => {
+                            warn!("Connection to Holochain app port {port} tried to send a message before authenticating. Dropping connection.");
+                            Err(())
+                        }
+                    }
+                }
+
+                warn!("Could not receive authentication message, the client either disconnected or sent a message that didn't decode to an authentication request. Dropping connection.");
+                Err(())
+            }).await;
+
+            match auth_payload_result {
+                Err(_) => {
+                    warn!("Connection to Holochain app port {port} timed out while waiting authenticating. Dropping connection.");
+                }
+                Ok(Err(_)) => {
+                    // Already logged, continue to drop connection
+                }
+                Ok(Ok(auth_payload)) => {
+                    let payload: AppAuthenticationRequest = SerializedBytes::from(
+                        holochain_serialized_bytes::UnsafeBytes::from(auth_payload),
+                    )
+                    .try_into()
+                    .unwrap();
```

Comment:
A malformed/malserialized auth payload will halt the conductor like this. Do you see a point to output an error instead and drop the connection?

Reply:
Hmm, this is a good point. It won't terminate the conductor because we're inside a Tokio task so it will just kill the task that is handling authentication and so the connection will die too which is desirable too. So the behaviour is 'correct' here but it's not obvious that is it. I'll update this so that:

- This is an `if let Err(_) = ... { warn!(...); return; }` which is more obviously a sensible way to handle the client mistake.
- There is a test checking that the conductor doesn't crash if you send a bad auth payload. I think think it will but we've just found out that sending bad admin/app payloads will kill interfaces so... worth checking!